#! /bin/sh /usr/share/dpatch/dpatch-run
## DP: fix for substitution of variables in 
## action e-mails, see:
## http://www.zabbix.com/forum/showthread.php?t=3969

@DPATCH@
--- expression.c	2006-09-13 14:19:23.000000000 +0200
+++ zabbix-1.1.2/src/zabbix_server/expression.c	2006-09-13 14:16:51.000000000 +0200
@@ -570,7 +570,12 @@
 #define MVAR_TRIGGER_SEVERITY		"{TRIGGER.SEVERITY}"
 
 #define STR_UNKNOWN_VARIAVLE		"*UNKNOWN*"
-static  void	substitute_simple_macros(DB_TRIGGER *trigger, DB_ACTION *action, char *data, int dala_max_len)
+
+#define MACRO_TYPE_TRIGGER_DESCRIPTION	1
+#define MACRO_TYPE_MESSAGE_SUBJECT	2
+#define MACRO_TYPE_MESSAGE_BODY		4
+
+static  void	substitute_simple_macros(DB_TRIGGER *trigger, DB_ACTION *action, char *data, int dala_max_len, int macro_type)
 {
 	char	sql[MAX_STRING_LEN];
 
@@ -604,13 +609,16 @@
 		snprintf(replace_to, sizeof(replace_to), "{");
 		var_len = 1;
 
-		if(strncmp(pr, MVAR_TRIGGER_NAME, strlen(MVAR_TRIGGER_NAME)) == 0)
+		if(macro_type & (MACRO_TYPE_MESSAGE_SUBJECT | MACRO_TYPE_MESSAGE_BODY) &&
+			strncmp(pr, MVAR_TRIGGER_NAME, strlen(MVAR_TRIGGER_NAME)) == 0)
 		{
 			var_len = strlen(MVAR_TRIGGER_NAME);
 
 			snprintf(replace_to, sizeof(replace_to), "%s", trigger->description);
+			substitute_simple_macros(trigger, action, replace_to, sizeof(replace_to), MACRO_TYPE_TRIGGER_DESCRIPTION);
 		}
-		else if(strncmp(pr, MVAR_HOST_NAME, strlen(MVAR_HOST_NAME)) == 0)
+		else if(macro_type & (MACRO_TYPE_MESSAGE_SUBJECT | MACRO_TYPE_MESSAGE_BODY | MACRO_TYPE_TRIGGER_DESCRIPTION) &&
+			strncmp(pr, MVAR_HOST_NAME, strlen(MVAR_HOST_NAME)) == 0)
 		{
 			var_len = strlen(MVAR_HOST_NAME);
 
@@ -634,7 +642,8 @@
 			}
 			DBfree_result(result);
 		}
-		else if(strncmp(pr, MVAR_TRIGGER_KEY, strlen(MVAR_TRIGGER_KEY)) == 0)
+		else if(macro_type & (MACRO_TYPE_MESSAGE_SUBJECT | MACRO_TYPE_MESSAGE_BODY) &&
+			strncmp(pr, MVAR_TRIGGER_KEY, strlen(MVAR_TRIGGER_KEY)) == 0)
 		{
 			var_len = strlen(MVAR_TRIGGER_KEY);
 
@@ -659,7 +668,8 @@
 
 			DBfree_result(result);
 		}
-		if(strncmp(pr, MVAR_IPADDRESS, strlen(MVAR_IPADDRESS)) == 0)
+		else if(macro_type & (MACRO_TYPE_MESSAGE_SUBJECT | MACRO_TYPE_MESSAGE_BODY) &&
+			strncmp(pr, MVAR_IPADDRESS, strlen(MVAR_IPADDRESS)) == 0)
 		{
 			var_len = strlen(MVAR_IPADDRESS);
 
@@ -683,7 +693,8 @@
 			}
 			DBfree_result(result);
 		}
-		else if(strncmp(pr, MVAR_DATE, strlen(MVAR_DATE)) == 0)
+		else if(macro_type & (MACRO_TYPE_MESSAGE_SUBJECT | MACRO_TYPE_MESSAGE_BODY) &&
+			strncmp(pr, MVAR_DATE, strlen(MVAR_DATE)) == 0)
 		{
 			var_len = strlen(MVAR_TIME);
 
@@ -691,7 +702,8 @@
 			tm	= localtime(&now);
 			snprintf(replace_to, sizeof(replace_to)-1, "%.4d.%.2d.%.2d", tm->tm_year+1900, tm->tm_mon+1, tm->tm_mday);
 		}
-		else if(strncmp(pr, MVAR_TIME, strlen(MVAR_TIME)) == 0)
+		else if(macro_type & (MACRO_TYPE_MESSAGE_SUBJECT | MACRO_TYPE_MESSAGE_BODY)&&
+			strncmp(pr, MVAR_TIME, strlen(MVAR_TIME)) == 0)
 		{
 			var_len = strlen(MVAR_TIME);
 
@@ -700,7 +712,8 @@
 			snprintf(replace_to, sizeof(replace_to), "%.2d:%.2d:%.2d",tm->tm_hour,tm->tm_min,tm->tm_sec);
 
 		}
-		else if(strncmp(pr, MVAR_TRIGGER_STATUS, strlen(MVAR_TRIGGER_STATUS)) == 0)
+		else if(macro_type & (MACRO_TYPE_MESSAGE_SUBJECT | MACRO_TYPE_MESSAGE_BODY) &&
+			strncmp(pr, MVAR_TRIGGER_STATUS, strlen(MVAR_TRIGGER_STATUS)) == 0)
 		{
 			/* NOTE: if you make changes for this bloc, don't forgot MVAR_TRIGGER_STATUS_OLD block */
 			var_len = strlen(MVAR_TRIGGER_STATUS);
@@ -710,7 +723,8 @@
 			else
 				snprintf(replace_to, sizeof(replace_to), "ON");
 		}
-		else if(strncmp(pr, MVAR_TRIGGER_STATUS_OLD, strlen(MVAR_TRIGGER_STATUS_OLD)) == 0)
+		else if(macro_type & (MACRO_TYPE_MESSAGE_SUBJECT | MACRO_TYPE_MESSAGE_BODY) && 
+			strncmp(pr, MVAR_TRIGGER_STATUS_OLD, strlen(MVAR_TRIGGER_STATUS_OLD)) == 0)
 		{
 			/* NOTE: if you make changes for this bloc, don't forgot MVAR_TRIGGER_STATUS block */
 			var_len = strlen(MVAR_TRIGGER_STATUS_OLD);
@@ -720,7 +734,8 @@
 			else
 				snprintf(replace_to, sizeof(replace_to), "ON");
 		}
-		else if(strncmp(pr, MVAR_TRIGGER_SEVERITY, strlen(MVAR_TRIGGER_SEVERITY)) == 0)
+		else if(macro_type & (MACRO_TYPE_MESSAGE_SUBJECT | MACRO_TYPE_MESSAGE_BODY) && 
+			strncmp(pr, MVAR_TRIGGER_SEVERITY, strlen(MVAR_TRIGGER_SEVERITY)) == 0)
 		{
 			var_len = strlen(MVAR_TRIGGER_SEVERITY);
 
@@ -786,7 +801,7 @@
 
 	zabbix_log(LOG_LEVEL_DEBUG, "In substitute_macros([%s])",data);
 
-	substitute_simple_macros(trigger, action, data, dala_max_len);
+	substitute_simple_macros(trigger, action, data, dala_max_len, MACRO_TYPE_MESSAGE_SUBJECT | MACRO_TYPE_MESSAGE_BODY);
 
 	*str_out = '\0';
 	outlen = sizeof(str_out) - 1;
